enhancements/etcd/storage-migration-for-etcd-encryption.md

```diff
@@ -0,0 +1,163 @@
+        for resource := range encryptedResources {
+            createStorageVersionMigration(resource)
+        }
+        waitUntilMigrationsComplete()
+    }
+}
+```
+
+## Risks and Mitigations
+
+* kube-storage-version-migrator is currently in alpha with a clear plan to beta.
+
+* If encryption keys change while api-servers are a different levels, forcing a storage migration might render some api-servers unable to read the etcd data. To mitigate, we don’t request storage migrations unless all control planes are at the same level.
+
+* In order to insure that storage migrations initiated by other components. For etcd encryption, for example, we distribute and maintain multiple read keys allowing for a storage migration to “occur” even when not all control planes have the latest shared write key.
+
+## Design Details
+
+#### How do operators communicate the need for migration?
+
+They create a `StorageVersionMigration` resource for the GroupVersionResource they wish to migrate.
+
+#### How do operators get progress info?
+
+ By examining `StorageVersionMigration.status.condition[@type='Succeeded|Running']`.
+
+#### What happens if multiple operators need migration?
+
+Each operator creates thier own request (`StorageVersionMigration`) for a specific GVR to be migrated. Each operator examines thier corresponding request for progress info. If multiple migration requests are made for the same GVR, those will be exected serially.
+
+#### Is migration cpu or memory bound on the migrator side or just heavy for etcd?
+
+>  TODO
+
+#### Do we need throttling?
+
+If we do, it would be a function of the upstream controller. Some throttling is currently done by the upstream controller.
+
+#### Do we want or need CRD migration as well which might mean to have a user-facing API.
+
+> TODO
+
+#### When do we trigger migration on upgrade?
+
+From an etcd encryption point of view, we do not trigger migration during an upgrade. Triggering a migration during an upgrade might cause an actual storage version migration that the control plane might not be ready for (e.g. some node are behind).
+
+#### Where do we store state of the last successful migration of a resource? We don't want to run a migration again if it was successful.
+
+The `StorageVersionMigration` resource can serve as a record of a successful migration.
+
+## Test Plan
+
+End-to-end tests will be implemented.
+
+## Graduation Criteria
+
+## Upgrade / Downgrade Strategy
+
+## Version Skew Strategy
+
+During an upgrade, we will always have skew among components, how will this impact your work?
+This component should not be invoked during an upgrade.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in Implementation History.
+
+## Drawbacks
+
+## Alternatives
+
+## Infrastructure Needed [optional]
+
+### New github projects:
+
+* `openshift/cluster-openshift-storage-migration-operator`
+* `openshift/kube-storage-version-migrator`
```

Comment:
Forks of upstream projects should have `kubernetes-` prefix.

Reply:
Updated